src/python/grpcio/grpc/framework/core/_end.py

```diff
@@ -0,0 +1,227 @@
+
+"""Implementation of base.End."""
+
+import abc
+import enum
+import threading
+import uuid
+
+from grpc.framework.core import _operation
+from grpc.framework.core import _utilities
+from grpc.framework.foundation import callable_util
+from grpc.framework.foundation import logging_pool
+from grpc.framework.interfaces.base import base
+from grpc.framework.interfaces.links import links
+from grpc.framework.interfaces.links import utilities
+
+_IDLE_ACTION_EXCEPTION_LOG_MESSAGE = 'Exception calling idle action!'
+
+
+class End(base.End, links.Link):
+  """A translation between base.End and links.Link."""
+  __metaclass__ = abc.ABCMeta
+
+
+@enum.unique
+class _State(enum.Enum):
+  STOPPED = 'stopped'
+  OPERATING = 'operating'
+  STOPPING = 'stopping'
+
+
+class _End(End):
+  """An End implementation."""
+
+  def __init__(self, servicer_package):
+    """Constructor.
+
+    Args:
+      servicer_package: A _ServicerPackage for servicing operations or None if
+        this end will not be used to service operations.
+    """
+    self._condition = threading.Condition()
+    self._servicer_package = servicer_package
+
+    self._mate = None
+
+    self._pool = None
+    self._state = _State.STOPPED
+    self._cycle_id = None
+
+    # Dictionary from operation IDs to ReceptionManager-or-None. A None value
+    # indicates an in-progress fire-and-forget operation for which the customer
+    # has chosen to ignore results.
+    self._operations = {}
+    self._stats = {outcome: 0 for outcome in base.Outcome}
+    self._idle_actions = []
+
+  def _termination_action(self, operation_id):
+    """Constructs the termination action for a single operation.
+
+    Args:
+      operation_id: An operation ID.
+
+    Returns:
+      A callable that takes an operation outcome for an argument to be used as
+        the termination action for the operation associated with the given
+        operation ID.
+    """
+    def termination_action(outcome):
+      with self._condition:
+        self._stats[outcome] += 1
+        self._operations.pop(operation_id, None)
+        if not self._operations:
+          for action in self._idle_actions:
+            self._pool.submit(action)
+          self._idle_actions = []
+          if self._state is _State.STOPPING:
+            self._condition.notify_all()
+    return termination_action
+
+  def _stop(self, immediate):
+    """Stops this End."""
+    with self._condition:
+      if self._state is _State.STOPPED:
+        return
+      elif self._state is _State.OPERATING:
+        if immediate:
+          for operation in self._operations.values():
+            operation.abort(base.Outcome.LOCAL_SHUTDOWN)
+        self._state = _State.STOPPING
+      cycle_id = self._cycle_id
+      while self._operations:
+        self._condition.wait()
+        if self._cycle_id is not cycle_id:
+          return
+      pool = self._pool
+
+    pool.shutdown(wait=True)
+
+    with self._condition:
```

Comment:
Yes, the ongoing operations depend on self because the End keeps track of which operations are active versus which have concluded.

I've changed up the lifecycle pretty dramatically now; please take another look.